src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/ConnectionImpl.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright The Original Author or Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+package io.jenkins.plugins.sshbuildagents.ssh.mina;
+
+import com.cloudbees.jenkins.plugins.sshcredentials.SSHAuthenticator;
+import com.cloudbees.plugins.credentials.common.StandardUsernameCredentials;
+import hudson.model.TaskListener;
+import io.jenkins.plugins.sshbuildagents.ssh.Connection;
+import io.jenkins.plugins.sshbuildagents.ssh.ServerHostKeyVerifier;
+import io.jenkins.plugins.sshbuildagents.ssh.ShellChannel;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.attribute.PosixFilePermission;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.sshd.client.SshClient;
+import org.apache.sshd.client.future.ConnectFuture;
+import org.apache.sshd.client.session.ClientSession;
+import org.apache.sshd.core.CoreModuleProperties;
+import org.apache.sshd.scp.client.DefaultScpClient;
+
+/**
+ * Implements {@link Connection} using the Apache Mina SSHD library
+ * https://github.com/apache/mina-sshd
+ *
+ */
+public class ConnectionImpl implements Connection {
+    /** The number of heartbeat packets lost before closing the connection. */
+    public static final int HEARTBEAT_MAX_RETRY = 6;
+
+    /** The size of the SSH window size in bytes. */
+    public static final long WINDOW_SIZE = 4L * 1024 * 1024;
```

Comment:
how did you come to those values? As they are very different from the default Mina values, maybe this would be nice to offer the possibility to override via `SystemProperties`

Reply:
Right so make it configurable can be a new feature in a follow up PR.